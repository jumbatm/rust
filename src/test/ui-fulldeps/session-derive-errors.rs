// check-fail
// Tests error conditions for specifying diagnostics using #[derive(AsSessionError)]

#![feature(rustc_private)]
#![crate_type = "lib"]

extern crate rustc_span;
use rustc_span::Span;

extern crate rustc_macros;
use rustc_macros::AsSessionError;

// The macro doesn't pull these crates in itself, because most internal use within the compiler is
// from contexts where referencing the crates is enough anyway.
extern crate rustc_errors;
extern crate rustc_session;

#[derive(AsSessionError)]
#[error = "Hello, world!"]
#[code = "E0123"]
struct Hello {}

#[derive(AsSessionError)]
#[error = "Hello, world!"]
#[code = "E0123"]
#[code = "E0456"] //~ ERROR Diagnostic ID multiply provided
struct ErrorSpecifiedTwice {}

// FIXME: Uncomment when emitting lints is supported.
/*
#[derive(AsSessionError)]
#[error = "Hello, world!"]
#[lint = "clashing_extern_declarations"]
#[lint = "improper_ctypes"] // FIXME: ERROR Diagnostic ID multiply provided
struct LintSpecifiedTwice {}
*/

#[derive(AsSessionError)]
#[code = "E0123"]
struct ErrorWithField {
    name: String,
    #[error = "This error has a field, and references {name}"]
    span: Span
}

#[derive(AsSessionError)]
#[code = "E0123"]
struct ErrorWithNonexistentField {
    #[error = "This error has a field, and references {name}"]
    //~^ ERROR no field `name` on this type
    span: Span
}

#[derive(AsSessionError)]
#[code = "E0123"]
#[error = "Something something"]
struct LabelOnSpan {
    #[label = "See here"]
    sp: Span
}

#[derive(AsSessionError)]
#[code = "E0123"]
#[error = "Something something"]
struct LabelOnNonSpan {
    #[label = "See here"]
    //~^ ERROR The `#[label = ...]` attribute can only be applied to fields of type Span
    id: u32,
}
