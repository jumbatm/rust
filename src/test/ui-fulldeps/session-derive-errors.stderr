error: `#[derive(SessionDiagnostic)]` can only be used on structs
  --> $DIR/session-derive-errors.rs:28:1
   |
LL | / #[code = "E0123"]
LL | |
LL | | enum SessionDiagnosticOnEnum {
LL | |     Foo,
LL | |     Bar,
LL | | }
   | |_^

error: `code` specified multiple times
  --> $DIR/session-derive-errors.rs:38:1
   |
LL | #[code = "E0456"]
   | ^^^^^^^^^^^^^^^^^

error: `code` not specified
  --> $DIR/session-derive-errors.rs:42:1
   |
LL | struct ErrorCodeNotProvided {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: use the [code = "..."] attribute to set this diagnostic's error code 

error: `name` doesn't refer to a field on this type
  --> $DIR/session-derive-errors.rs:64:5
   |
LL |     #[error = "This error has a field, and references {name}"]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: invalid format string: expected `'}'` but string was terminated
  --> $DIR/session-derive-errors.rs:71:2
   |
LL | #[code = "E0123"]
   |               - because of this opening brace
LL | #[error = "This is missing a closing brace: {name"]
   |  ^ expected `'}'` in format string
   |
   = note: if you intended to print `{`, you can escape it using `{{`
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: invalid format string: unmatched `}` found
  --> $DIR/session-derive-errors.rs:80:2
   |
LL | #[error = "This is missing an opening brace: name}"]
   |  ^ unmatched `}` in format string
   |
   = note: if you intended to print `}`, you can escape it using `}}`
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: The `#[label = ...]` attribute can only be applied to fields of type Span
  --> $DIR/session-derive-errors.rs:99:5
   |
LL |     #[label = "See here"]
   |     ^^^^^^^^^^^^^^^^^^^^^

error: `nonsense` is not a valid key for `#[suggestion(...)]`
  --> $DIR/session-derive-errors.rs:124:18
   |
LL |     #[suggestion(nonsense = "This is nonsense")]
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: `msg` is not a valid key for `#[suggestion(...)]`
  --> $DIR/session-derive-errors.rs:132:18
   |
LL |     #[suggestion(msg = "This is a suggestion")]
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: missing suggestion message
  --> $DIR/session-derive-errors.rs:140:7
   |
LL |     #[suggestion(code = "This is suggested code")]
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: provide a suggestion message using #[suggestion(message = "...")]

error: wrong types for suggestion
  --> $DIR/session-derive-errors.rs:155:5
   |
LL | /     #[suggestion(message = "This is a message", code = "This is suggested code")]
LL | |
LL | |     suggestion: Applicability,
   | |_____________________________^
   |
   = help: #[suggestion(...)] should be applied to fields of type (Span, Applicability)

error: wrong types for suggestion
  --> $DIR/session-derive-errors.rs:163:5
   |
LL | /     #[suggestion(message = "This is a message", code = "This is suggested code")]
LL | |
LL | |     suggestion: Span,
   | |____________________^
   |
   = help: #[suggestion(...)] should be applied to fields of type (Span, Applicability)

error: type of field annotated with `#[suggestion(...)]` contains more than one Span
  --> $DIR/session-derive-errors.rs:171:5
   |
LL | /     #[suggestion(message = "This is a message", code = "This is suggested code")]
LL | |
LL | |     suggestion: (Span, Span, Applicability),
   | |___________________________________________^

error: type of field annotated with `#[suggestion(...)]` contains more than one Applicability
  --> $DIR/session-derive-errors.rs:179:5
   |
LL | /     #[suggestion(message = "This is a message", code = "This is suggested code")]
LL | |
LL | |     suggestion: (Applicability, Applicability, Span),
   | |____________________________________________________^

error: invalid annotation list `#[label(...)]`
  --> $DIR/session-derive-errors.rs:187:7
   |
LL |     #[label("wrong kind of annotation for label")]
   |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 15 previous errors

